### ISSUE TITLE
Possible bug with either websockets or access to widget in page

### ISSUE THREAD

**Issue by amckay1** [JuliaWeb/Mux.jl#80]
This popped up on me recently and haven't been able to figure out what went wrong. No bites on discourse so might be a real issue. As the example code should replicate, when accessing a widget from "/" it is responsive and "clicked" is printed in the REPL, while "/page1" is unresponsive for the same required widget. Not the behavior I was expecting or experiencing previously.

I do get a websockets error (pasted below) but unclear why it would selectively affect the "/page1" endpoint. Hopefully this is an obvious fix, but I was able to replicate the issue locally and on an aws server, both running Julia 1.0.2, Mux v0.5.3 and Interact v0.9.0. Thinking this is a Mux issue and not an interact issue because of the websocket error and the widget working at least for the "/" endpoint.

Thank you for any help you can provide!
Andrew

Error:
```
Error handling websocket connection: WebSockets.WebSocketClosedError(“ws|server respond to OPCODE_CLOSE 1001:Going Away”)"
```

Example code:
```
using Mux, Interact                                                                                                                                                       
                                                                                                                                                                                  
 # setting up Interact.jl widget                                                                                                                                                  
 b = button("Test click")                                                                                                                                                         
 on(n -> println("clicked"), b)                                                                                                                                        
                                                                                                                                                                                  
 # putting together apps for root and id1 page using same widget                                                                                                                  
 app1 = page("/", req -> b)                                                                                                                                                    
 app2 = page("/page1", req -> b)                                                                                                                                                       
                                                                                                                                                                                  
 # stacking together                                                                                                                                                              
 app = Mux.stack(app1, app2)                                                                                                                                                 
                                                                                                                                                                                  
 # serving                                                                                                                                                                        
 WebIO.webio_serve(app, 8000) 
```

**Comment by amckay1** [comment]
Huh, thanks for the ideas, decided to check the stuck server side coroutine idea with a fresh restart but to no avail: same behavior, same "Error handling websocket connection" error message. Not sure if fully ruled out, but now betting the websockets error is the underlying issue, going to try digging into that more, will swing back here if I make any headway. Thanks again!

**Comment by hustf** [comment]
I think [this](https://github.com/JuliaGizmos/WebIO.jl/blob/ef95ce36851b679e0e0ba778cc8f715307e5c5da/src/providers/mux.jl) at line 38 is the server side code which could benfit from an upgrade from read to readguarded. I suggest you open an issue on WebIO and then follow up with a pull request. Tag me if you would like a review.

I'm able to reproduce just as you say. I used Chrome to inspect the first page, button worked. Upon switching to the second page, the first one naturally closed down and produced the 'Error'. Which should be of no consequence at all since nobody cares what the coroutine does anymore. A new task/ coroutine is generated every time you load a new page.

```
function create_socket(req)
    sock = req[:socket]
    conn = WebSockConnection(sock)

    t = @async while isopen(sock)
        data = read(sock)

        msg = JSON.parse(String(data))
        WebIO.dispatch(conn, msg)
    end

    wait(t)
end
```

**Comment by amckay1** [comment]
Thanks hustf! As soon as the holiday season lightens up on me I'll take a closer look and open an issue with WebIO, probably tagging you in the pull request and/or the issue to be sure I clarified the issue correctly.

So if I'm understanding you right, that WebIO issue is likely separate from the behavior I described that appears to be a bug? Also good to know, with this lead perhaps I can make more headway on tracking down the bug.

Thanks again and happy holidays!

**Comment by hustf** [comment]
I was actually not reading properly. I can't reproduce the non-responsive behaviour. I get the error message whenever I close a page, but I can load both localhost:8000 and localhost:8000/page1. If I have several pages open at the same time and close one, the other ones are not affected. Perhaps you can report in more detail exactly what you do? Operating system, and which front-end?

**Comment by amckay1** [comment]
Hey hustf, thanks for looking at this again, once you couldn't reproduce the non-responsive behavior I took another look at it. Remembering your comment on a potential WebIO issue, I found my WebIO was at 0.6.1 and I was able to update it to 0.7.0, and that *seems* to have fixed the issue?

I'm only a little mystified because I thought I had updated everything, and I don't see how I would have selectively updated Mux + Interact but not WebIO, resulting in some incompatibility, but that could be the case. For the record, I have a macbook running OSX 10.14 and Julia 1.02, and I was using Chrome.

Regardless, I better close this issue for now since it seems to be resolved with the combination of Interact.jl v0.9.0, Mux.jl v0.5.3, and WebIO v0.7.0. Thank you again! -Andrew
